includes/sanitizers/class-amp-style-sanitizer.php

```diff
@@ -349,230 +431,741 @@ private function process_link_element( DOMElement $element ) {
+					}
+				}
+			}
+
+			/**
+			 * Source URL lists.
+			 *
+			 * @var URL[] $source_file_urls
+			 * @var URL[] $source_data_urls
+			 */
+			$source_file_urls = array();
+			$source_data_urls = array();
+			foreach ( $sources as $i => $source ) {
+				if ( $source[0] instanceof URL ) {
+					if ( 'data:' === substr( $source[0]->getURL()->getString(), 0, 5 ) ) {
+						$source_data_urls[ $i ] = $source[0];
+					} else {
+						$source_file_urls[ $i ] = $source[0];
+					}
+				}
+			}
+
+			// Convert data: URLs into regular URLs, assuming there will be a file present (e.g. woff fonts in core themes).
+			if ( empty( $source_file_urls ) ) {
 				continue;
 			}
+			$source_file_url = current( $source_file_urls );
+			foreach ( $source_data_urls as $i => $data_url ) {
+				$mime_type = strtok( substr( $data_url->getURL()->getString(), 5 ), ';' );
+				if ( ! $mime_type ) {
+					continue;
+				}
+				$extension   = preg_replace( ':.+/(.+-)?:', '', $mime_type );
+				$guessed_url = preg_replace(
+					':(?<=\.)\w+(\?.*)?(#.*)?$:', // Match the file extension in the URL.
+					$extension,
+					$source_file_url->getURL()->getString(),
+					1,
+					$count
+				);
+				if ( $count ) {
+					$data_url->getURL()->setString( $guessed_url );
+				}
+			}
+		}
+	}
+
+	/**
+	 * Process CSS keyframes.
+	 *
+	 * @since 1.0
+	 * @link https://www.ampproject.org/docs/design/responsive/style_pages#restricted-styles.
+	 * @link https://github.com/ampproject/amphtml/blob/b685a0780a7f59313666225478b2b79b463bcd0b/validator/validator-main.protoascii#L1002-L1043
+	 * @todo Tree shaking could be extended to keyframes, to omit a keyframe if it is not referenced by any rule.
+	 *
+	 * @param KeyFrame $css_list Ruleset.
+	 * @param array    $options  Options.
+	 * @return array Validation errors.
+	 */
+	private function process_css_keyframes( KeyFrame $css_list, $options ) {
+		$validation_errors = array();
+		if ( ! empty( $options['property_whitelist'] ) ) {
+			foreach ( $css_list->getContents() as $rules ) {
+				if ( ! ( $rules instanceof DeclarationBlock ) ) {
+					$validation_errors[] = array(
+						'code'    => 'unrecognized_css',
+						'message' => __( 'Unrecognized CSS removed.', 'amp' ),
+					);
+					$css_list->remove( $rules );
+					continue;
+				}
+
+				$validation_errors = array_merge(
+					$validation_errors,
+					$this->transform_important_qualifiers( $rules, $css_list )
+				);
+
+				$properties = $rules->getRules();
+				foreach ( $properties as $property ) {
+					$vendorless_property_name = preg_replace( '/^-\w+-/', '', $property->getRule() );
+					if ( ! in_array( $vendorless_property_name, $options['property_whitelist'], true ) ) {
+						$validation_errors[] = array(
+							'code'           => 'illegal_css_property',
+							'property_name'  => $property->getRule(),
+							'property_value' => $property->getValue(),
+						);
+						$rules->removeRule( $property->getRule() );
```

Comment:
This removal should be removed since it will break many keyframe animations out there which don't use hardware accelerated properties.